redshift/redshift.go

```diff
@@ -7,113 +7,180 @@ import (
-	pwd                = flag.String("redshiftpassword", "", "Password for the redshift user")
-	timeout            = flag.Duration("redshiftconnecttimeout", 10*time.Second,
-		"Timeout while connecting to Redshift. Defaults to 10 seconds.")
 )
 
-// NewRedshift returns a pointer to a new redshift object using configuration values set in the
-// flags.
-func NewRedshift() (*Redshift, error) {
+// NewRedshift returns a pointer to a new redshift object using configuration values passed in
+// on instantiation and the AWS env vars we assume exist
+// Don't need to pass s3 info unless doing a COPY operation
+func NewRedshift(host, port, db, user, password string, timeout int, s3Info S3Info) (*Redshift, error) {
 	flag.Parse()
-	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", *host, *port, *db, int(timeout.Seconds()))
-	log.Println("Connecting to Reshift Source: ", source)
-	source += fmt.Sprintf(" user=%s password=%s", *user, *pwd)
+	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", host, port, db, timeout)
+	log.Println("Connecting to Redshift Source: ", source)
+	source += fmt.Sprintf(" user=%s password=%s", user, password)
 	sqldb, err := sql.Open("postgres", source)
 	if err != nil {
 		return nil, err
 	}
-	return &Redshift{sqldb, awsAccessKeyID, awsSecretAccessKey}, nil
+	return &Redshift{sqldb, s3Info}, nil
 }
 
-func (r *Redshift) logAndExec(cmd string, creds bool) (sql.Result, error) {
-	log.Print("Executing Redshift command: ", cmd)
-	if creds {
-		cmd += fmt.Sprintf(" CREDENTIALS '%s=%s;%s=%s'", "aws_access_key_id", r.accessID, "aws_secret_access_key", r.secretKey)
-	}
+func (r *Redshift) logAndExec(cmd string) (sql.Result, error) {
+	log.Printf("Executing Redshift command: %s", cmd)
 	return r.Exec(cmd)
 }
 
-// CopyJSONDataFromS3 copies JSON data present in an S3 file into a redshift table.
-func (r *Redshift) CopyJSONDataFromS3(schema, table, file, jsonpathsFile, awsRegion string) error {
-	copyCmd := fmt.Sprintf(
-		`COPY "%s"."%s" FROM '%s' WITH json '%s' region '%s' timeformat 'epochsecs' COMPUPDATE ON`,
-		schema, table, file, jsonpathsFile, awsRegion)
-	_, err := r.logAndExec(copyCmd, true)
+// RunJSONCopy copies JSON data present in an S3 file into a redshift table.
+// this is meant to be run in a transaction, so the first arg must be a sql.Tx
+// if not using jsonPaths, set to "auto"
+func (r *Redshift) RunJSONCopy(tx *sql.Tx, schema, table, filename, jsonPaths string, creds, gzip bool) error {
+	var credSQL string
+	var credArgs []interface{}
+	if creds {
+		credSQL = `CREDENTIALS 'aws_access_key_id=?;aws_secret_access_key=?'`
+		credArgs = []interface{}{r.s3Info.AccessID, r.s3Info.SecretKey}
+	}
+	gzipSQL := ""
+	if gzip {
+		gzipSQL = "GZIP"
+	}
+	copySQL := `COPY "?"."?" FROM '?' WITH ? JSON '?' REGION '?' TIMEFORMAT 'auto' STATUPDATE COMPUPDATE ON %s`
+	copyStmt, err := r.Prepare(fmt.Sprintf(copySQL, credSQL))
```

Comment:
Could we move the prepared queries outside the function? From what I understand, that's one of the major advantages of using prepared queries because we can re-use the same ones with different arguments. Also saves us from doing an extra hop on every invocation but I think that's somewhat minor advantage for this case.